### PR TITLE
Fixes #19868: Rudder package should accept plugins with same minor version, not just patch version

### PR DESCRIPTION
**Pull request by VinceMacBuche**
https://issues.rudder.io/issues/19868

### CODE REVIEW

**Comment by VinceMacBuche** [comment]
Commit modified

File: relay/sources/rudder-pkg/lib/rudder-pkg/rudderPkgUtils.py [hunk 1]
```diff
@@ -356,28 +356,28 @@ def package_check(metadata, version):
...
+
+def check_exact_compatibility(plugin_version):
+    # check that the given version is exactly the same as Rudder one
+    rudder_version = plugin_version.group(1).replace('~alpha', 'a').replace('~beta', 'b').replace('~rc', 'c')
+    return rudder_version == RUDDER_VERSION
+
+def check_loose_compatibility(plugin_version):
+    # check that the given version is compatible with Rudder minor one
+    rudder_minor = plugin_version.group(2)
+    return rudder_minor == RUDDER_MINOR
```
Comment (fanf) [inline relay/sources/rudder-pkg/lib/rudder-pkg/rudderPkgUtils.py:375]: I had a minute of "wait, does that means it doesn't compare major version", but it's actually ok. Perhaps update the commen to `... with Rudder major.minor one` to better convey the intent

File: relay/sources/rudder-pkg/lib/rudder-pkg/rudderPkgUtils.py [hunk 2]
```diff
@@ -660,9 +660,10 @@ def list_plugin_name():
                 elif m.group(1) == 'rc':
                     step = 'c'
                 step += m.group(2)
-            m = re.match(r'main_version *= *((\d+\.\d+)\.\d+)', line)
+            m = re.match(r'main_version *= *((\d+\.\d+)\.(\d+))', line)
             if m:
-                RUDDER_MAJOR = m.group(2)
+                RUDDER_MINOR = m.group(2)
```
Comment (fanf) [inline relay/sources/rudder-pkg/lib/rudder-pkg/rudderPkgUtils.py:665]: perhaps had a comment with `# RUDDER_MINOR is "x.y" in "x.y.z" version pattern`

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

File: relay/sources/rudder-pkg/lib/rudder-pkg/rudderPkgUtils.py [hunk 3]
```diff
@@ -351,33 +350,41 @@ def package_check(metadata, version):
...
+    rudder_minor = plugin_version.group(2)
+    return rudder_minor == RUDDER_MINOR
+
+
+def check_plugin_compatibility(metadata, exact_version):
+    # check that the given version is compatible with Rudder one (loose or exact)
+    plugin_version = extract_plugin_version(metadata)
+    if exact_version:
+        return check_exact_compatibility(plugin_version)
+    return check_exact_compatibility(plugin_version) or check_loose_compatibility(metadata)
```
Comment (Fdall) [inline relay/sources/rudder-pkg/lib/rudder-pkg/rudderPkgUtils.py:387]:
```suggestion
    return check_exact_compatibility(plugin_version) or check_loose_compatibility(plugin_version)
```

File: relay/sources/rudder-pkg/lib/rudder-pkg/rpkg.py [hunk 4]
```diff
@@ -50,7 +52,7 @@ def __ge__(self, other):
 class PluginVersion:
     def __init__(self, pluginLongVersion):
         match = re.search(
-            r'(?P<rudderVersion>(?P<rudderMajor>[0-9]+\.[0-9]+)(\.[0-9]+(~(beta|rc)[0-9]+)?)?)-(?P<pluginShortVersion>[0-9]+\.[0-9]+)(-(?P<mode>[a-zA-Z]+))?',
+            r'(?P<rudderVersion>(?P<rudderMajor>[0-9]+\.[0-9]+)(\.[0-9]+(~(alpha|beta|rc)[0-9]+)?)?)-(?P<pluginShortVersion>[0-9]+\.[0-9]+)(-(?P<mode>[a-zA-Z]+))?',
```
Comment (Fdall) [inline relay/sources/rudder-pkg/lib/rudder-pkg/rpkg.py:55]: We should update the version comparison tests to match the new plugin version format.

Comment (VinceMacBuche) [reply relay/sources/rudder-pkg/lib/rudder-pkg/rpkg.py:55]: damn i forgot one, good catch

**Comment by VinceMacBuche** [comment]
Commit modified

**Comment by VinceMacBuche** [comment]
Commit modified

File: relay/sources/rudder-pkg/lib/rudder-pkg/rpkg.py (hunk 4 above)
Comment (VinceMacBuche) [reply relay/sources/rudder-pkg/lib/rudder-pkg/rpkg.py:55]: I updated some tests! Is is ok for you now @Fdall ?

**Review by Fdall — APPROVED** [review]

**Pull request merged**